File: Programmers/2021_카카오_채용연계형_인턴십/거리두기 확인하기.py
from collections import deque


def bfs(p, x, y, d):
    di = [[-1, 0], [1, 0], [0, -1], [0, 1]]
    visited = [[0] * 5 for _ in range(5)]

    q = deque([[x, y, d]])

    while q:
        x, y, d = q.popleft()
        visited[x][y] = 1

        for i in range(4):
            nx = x + di[i][0]
            ny = y + di[i][1]
            nd = d + 1

            if 0 <= nx < 5 and 0 <= ny < 5 and not visited[nx][ny]:
                visited[nx][ny] = 1

                if p[nx][ny] == 'P':
                    if nd <= 2:
                        return False

                elif p[nx][ny] == 'O':
                    if nd == 1:
                        q.append([nx, ny, nd])

    return True


def solution(places):
    answer = []

    for p in places:
        flag = 1

        for i in range(5):
            for j in range(5):
                if p[i][j] == 'P':
                    result = bfs(p, i, j, 0)
                    if not result:
                        flag = 0

        answer.append(flag)

    return answer

File: Programmers/2021_카카오_채용연계형_인턴십/거리두기 확인하기.js
function solution(places) {
    var answer = [];
    for (const p of places) {
        let flag = 1;

        for (let i = 0; i < 5; i++) {
            for (let j = 0; j < 5; j++) {
                if (flag) {
                    if (p[i][j] == 'P') {
                        let result = bfs(p, i, j, 0);
                        if (!result) {
                            flag = 0;
                        }
                    }
                }
            }
        }
        answer.push(flag);
    }


    return answer;
}

function bfs(p, x, y, d) {

    const di = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    const visited = Array.from({ length: 5 }, () => Array(5).fill(0));

    const q = [[x, y, d]];

    while (q.length > 0) {
        const [x, y, d] = q.shift();
        visited[x][y] = 1;

        for (let i = 0; i < 4; i++) {
            const nx = x + di[i][0];
            const ny = y + di[i][1];
            const nd = d + 1;

            if (0 <= nx && nx < 5 && 0 <= ny && ny < 5 && visited[nx][ny] == 0) {
                visited[nx][ny] = 1;

                if (p[nx][ny] == 'P') {
                    if (nd <= 2) {
                        return false;
                    }
                } else if (p[nx][ny] == 'O') {
                    if (nd < 2) {
                        q.push([nx, ny, nd]);
                    }
                }
            }
        }
    }
    return true;


}